ext/mbstring/mbstring.c

```diff
@@ -1544,61 +1544,58 @@ static inline int php_mb_check_code_point(zend_long cp)
+	zend_bool substitute_is_null = 1;
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "|z", &arg1) == FAILURE) {
-		RETURN_THROWS();
-	}
+	ZEND_PARSE_PARAMETERS_START(0, 1)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_STR_OR_LONG_OR_NULL(substitute_character, substitute_codepoint, substitute_is_null)
+	ZEND_PARSE_PARAMETERS_END();
 
-	if (!arg1) {
+	if (substitute_is_null) {
 		if (MBSTRG(current_filter_illegal_mode) == MBFL_OUTPUTFILTER_ILLEGAL_MODE_NONE) {
 			RETURN_STRING("none");
-		} else if (MBSTRG(current_filter_illegal_mode) == MBFL_OUTPUTFILTER_ILLEGAL_MODE_LONG) {
+		}
+		if (MBSTRG(current_filter_illegal_mode) == MBFL_OUTPUTFILTER_ILLEGAL_MODE_LONG) {
 			RETURN_STRING("long");
-		} else if (MBSTRG(current_filter_illegal_mode) == MBFL_OUTPUTFILTER_ILLEGAL_MODE_ENTITY) {
+		}
+		if (MBSTRG(current_filter_illegal_mode) == MBFL_OUTPUTFILTER_ILLEGAL_MODE_ENTITY) {
 			RETURN_STRING("entity");
-		} else {
-			RETURN_LONG(MBSTRG(current_filter_illegal_substchar));
 		}
-	} else {
-		RETVAL_TRUE;
-
-		switch (Z_TYPE_P(arg1)) {
-			case IS_STRING:
-				if (strncasecmp("none", Z_STRVAL_P(arg1), Z_STRLEN_P(arg1)) == 0) {
-					MBSTRG(current_filter_illegal_mode) = MBFL_OUTPUTFILTER_ILLEGAL_MODE_NONE;
-				} else if (strncasecmp("long", Z_STRVAL_P(arg1), Z_STRLEN_P(arg1)) == 0) {
-					MBSTRG(current_filter_illegal_mode) = MBFL_OUTPUTFILTER_ILLEGAL_MODE_LONG;
-				} else if (strncasecmp("entity", Z_STRVAL_P(arg1), Z_STRLEN_P(arg1)) == 0) {
-					MBSTRG(current_filter_illegal_mode) = MBFL_OUTPUTFILTER_ILLEGAL_MODE_ENTITY;
-				} else {
-					convert_to_long_ex(arg1);
+		RETURN_LONG(MBSTRG(current_filter_illegal_substchar));
+	}
 
-					if (php_mb_check_code_point(Z_LVAL_P(arg1))) {
-						MBSTRG(current_filter_illegal_mode) = MBFL_OUTPUTFILTER_ILLEGAL_MODE_CHAR;
-						MBSTRG(current_filter_illegal_substchar) = Z_LVAL_P(arg1);
-					} else {
-						php_error_docref(NULL, E_WARNING, "Unknown character");
-						RETURN_FALSE;
-					}
-				}
-				break;
-			default:
-				convert_to_long_ex(arg1);
-				if (php_mb_check_code_point(Z_LVAL_P(arg1))) {
-					MBSTRG(current_filter_illegal_mode) = MBFL_OUTPUTFILTER_ILLEGAL_MODE_CHAR;
-					MBSTRG(current_filter_illegal_substchar) = Z_LVAL_P(arg1);
-				} else {
-					php_error_docref(NULL, E_WARNING, "Unknown character");
-					RETURN_FALSE;
-				}
-				break;
+	if (substitute_character != NULL) {
+		if (zend_string_equals_literal(substitute_character, "none")) {
+			MBSTRG(current_filter_illegal_mode) = MBFL_OUTPUTFILTER_ILLEGAL_MODE_NONE;
+			RETURN_TRUE;
+		}
+		if (zend_string_equals_literal(substitute_character, "long")) {
+			MBSTRG(current_filter_illegal_mode) = MBFL_OUTPUTFILTER_ILLEGAL_MODE_LONG;
+			RETURN_TRUE;
+		}
+		if (zend_string_equals_literal(substitute_character, "entity")) {
+			MBSTRG(current_filter_illegal_mode) = MBFL_OUTPUTFILTER_ILLEGAL_MODE_ENTITY;
+			RETURN_TRUE;
 		}
+		/* Invalid string value */
+		zend_argument_value_error(1, "must be 'none', 'long' or 'entity'");
```

Comment:
```suggestion
		zend_argument_value_error(1, "must be 'none', 'long', 'entity' or a valid codepoint");
```

Maybe? As is, it kinda sounds like you can *only* pass in those strings.

Reply:
Indeed, adjusted the message accordingly.